stackslib/src/clarity_vm/tests/contracts.rs

```diff
@@ -1114,3 +1114,284 @@ fn test_block_heights() {
+                    &contract_e2c1_2,
+                    ASTRules::PrecheckSize,
+                )
+                .unwrap();
+            clarity_db
+                .save_analysis(&contract_id_e2c2, &analysis)
+                .unwrap();
+
+            // Publish the Clarity 2 contract
+            clarity_db
+                .initialize_smart_contract(
+                    &contract_id_e2c2,
+                    ClarityVersion::Clarity2,
+                    &ast,
+                    contract_e2c1_2,
+                    None,
+                    |_, _| false,
+                )
+                .unwrap();
+        });
+    });
+
+    // Advance to epoch 3
+    while sim.block_height <= 7 {
+        sim.execute_next_block(|_env| {});
+    }
+
+    // Deploy the Clarity 3 contract in the next block
+    sim.execute_next_block_as_conn(|conn| {
+        conn.as_transaction(|clarity_db| {
+            // Analyze the Clarity 3 contract
+            let (ast, analysis) = clarity_db
+                .analyze_smart_contract(
+                    &contract_id_e3c3,
+                    ClarityVersion::Clarity3,
+                    &contract_e3c3,
+                    ASTRules::PrecheckSize,
+                )
+                .unwrap();
+
+            // Publish the Clarity 3 contract
+            clarity_db
+                .initialize_smart_contract(
+                    &contract_id_e3c3,
+                    ClarityVersion::Clarity3,
+                    &ast,
+                    &contract_e3c3,
+                    None,
+                    |_, _| false,
+                )
+                .unwrap();
+        });
+    });
+
+    // Call the Clarity 3 contract and validate the results
+    sim.execute_next_block_as_conn(|conn| {
+        let mut tx = conn.start_transaction_processing();
+        assert_eq!(
+            Value::Int(124),
+            tx.eval_read_only(&contract_id_e3c3, "(call-e2 1)").unwrap()
+        );
+        assert_eq!(
+            Value::Int(456),
+            tx.eval_read_only(&contract_id_e3c3, "(call-e2 2)").unwrap()
+        );
+    });
+}
+
+/// Test passing a Clarity 3 contract using `stacks-block-height` and
+/// `tenure-height` as a trait into a Clarity 1 and Clarity 2 contract.
+#[test]
+fn test_block_heights_across_versions_traits() {
+    let mut sim = ClarityTestSim::new();
+    sim.epoch_bounds = vec![0, 1, 2, 3, 4, 5, 6, 7];
+
+    let contract_id_e2c1 = QualifiedContractIdentifier::local("epoch-2-clarity-1").unwrap();
+    let contract_id_e2c2 = QualifiedContractIdentifier::local("epoch-2-clarity-2").unwrap();
+    let contract_id_e3c3 = QualifiedContractIdentifier::local("epoch-3-clarity-3").unwrap();
+
+    let contract_e2c1_2 = r#"
+        (define-trait getter ((get-int () (response uint uint))))
+        (define-public (get-it (get-trait <getter>))
+            (contract-call? get-trait get-int)
+        )
+        "#;
+    let contract_e3c3 = format!(
```

Comment:
In a similar vein, what happens if the trait implementation was deployed in epoch 2.x, and had a bound name `stacks-block-height` (or `tenure-height`), and was called in epoch 3.x?  The VM should continue to treat the epoch 2.x's bound names as if they were variables, and not reserved words, but I just want to see a test that confirms this (it may exist; I just may not have gotten to it yet :))

Reply:
Added `test_block_heights_across_versions_traits_2_from_3`